src/tracker/inflights.rs

```diff
@@ -37,13 +42,43 @@ impl Inflights {
             start: 0,
             count: 0,
             cap,
+            incoming_cap: None,
+        }
+    }
+
+    /// Adjust inflight buffer capacity. Set it to `0` will disable the progress.
+    // Calling it between `self.full()` and `self.add()` can cause a panic.
+    pub fn set_cap(&mut self, incoming_cap: usize) {
+        match self.cap.cmp(&incoming_cap) {
+            Ordering::Equal => self.incoming_cap = None,
+            Ordering::Less => {
+                self.buffer.reserve(incoming_cap - self.count);
+                self.cap = incoming_cap;
+                self.incoming_cap = None;
+            }
+            Ordering::Greater => {
+                if self.count <= incoming_cap && self.start + self.count < incoming_cap {
+                    self.cap = incoming_cap;
+                    let (cur_cap, cur_len) = (self.buffer.capacity(), self.buffer.len());
+                    if cur_cap > incoming_cap && cur_len <= incoming_cap {
+                        // TODO: Simplify it after `shrink_to` is stable.
+                        unsafe {
+                            self.buffer.set_len(incoming_cap);
```

Comment:
I don't think it's a good idea. It will introduce read uninitialized undefined behavior.